internal/state/execution.go

```diff
@@ -433,35 +435,49 @@ func buildLastExtendedCommitInfo(ec *types.ExtendedCommit, store Store, initialH
 		panic(fmt.Errorf("failed to load validator set at height %d, initial height %d: %w", ec.Height, initialHeight, err))
 	}
 
-	vs := make([]abci.ExtendedVoteInfo, ec.Size())
-	for i, ecs := range ec.ExtendedSignatures {
+	var (
+		ecSize    = ec.Size()
+		valSetLen = len(valSet.Validators)
+	)
+
+	// Ensure that the size of the validator set in the extended commit matches
+	// the size of the validator set in the state store.
+	if ecSize != valSetLen {
+		panic(fmt.Errorf(
+			"extended commit size (%d) does not match validator set length (%d) at height %d\n\n%v\n\n%v",
+			ecSize, valSetLen, ec.Height, ec.ExtendedSignatures, valSet.Validators,
+		))
+	}
+
+	votes := make([]abci.ExtendedVoteInfo, ecSize)
+	for i, val := range valSet.Validators {
+		ecs := ec.ExtendedSignatures[i]
+
+		// Absent signatures have empty validator addresses, but otherwise we
+		// expect the validator addresses to be the same.
+		if ecs.BlockIDFlag != types.BlockIDFlagAbsent && !bytes.Equal(ecs.ValidatorAddress, val.Address) {
+			panic(fmt.Errorf("validator address of extended commit signature in position %d (%s) does not match the corresponding validator's at height %d (%s)",
+				i, ecs.ValidatorAddress, ec.Height, val.Address,
+			))
+		}
+
 		var ext []byte
 		if ecs.BlockIDFlag == types.BlockIDFlagCommit {
 			// We only care about vote extensions if a validator has voted to
 			// commit.
 			ext = ecs.Extension
 		}
```

Comment:
As I mentioned in #8487, we should (at least) log an error here if the precommit is nil, and it has an extension.
Or maybe, it's better to write that in `ValidateBasic` so we check that more generally.

Reply:
@williambanfield, thinking of metrics, do you think it would be worth adding a metric like "`nil` precommits received with an extension"... as that'd be evidence of (at least) a bug in the App (if not outright byzantine behavior).
Of course, I'm not suggesting to do this in this PR :-)

Reply:
I was just thinking of adding this check to `ValidateBasic`. Technically if we get a nil precommit with vote extensions that's a violation of the spec.

Reply:
I think so, thanks @thanethomson !

Reply:
> thinking of metrics, do you think it would be worth adding a metric like "nil precommits received with an extension"... as that'd be evidence of (at least) a bug in the App


Good question, opened https://github.com/tendermint/tendermint/issues/8498 to discuss.